server/functions/0_parameters.go

```diff
@@ -0,0 +1,190 @@
+	ParameterType_Timestamp                      // The parameter is a Timestamp type
+
+	ParameterType_Length // The number of parameters. This should always be last in the enum declaration.
+)
+
+// ptCasts contains an array of all potential casts for each parameter type
+var ptCasts = [ParameterType_Length][]ParameterType{
+	{ParameterType_Integer, ParameterType_Float, ParameterType_Numeric, ParameterType_String, ParameterType_Timestamp}, // ParameterType_Null
+	{ParameterType_Float, ParameterType_Numeric, ParameterType_String},                                                 // ParameterType_Integer
+	{ParameterType_Numeric, ParameterType_String},                                                                      // ParameterType_Float
+	{ParameterType_String}, // ParameterType_Numeric
+	{ParameterType_Integer, ParameterType_Float, ParameterType_Numeric, ParameterType_Timestamp}, // ParameterType_String
+	{}, // ParameterType_Timestamp
+}
+
+// PotentialCasts returns all potential casts for the current type. For example, an Integer may be cast to a Float.
+// Casts may be bidirectional, as a String may cast to an Integer, and an Integer may cast to a String.
+func (pt ParameterType) PotentialCasts() []ParameterType {
+	return ptCasts[pt]
+}
+
+// PotentialCasts returns all potential casts for the current type. For example, an Integer may be cast to a Float.
+// Casts may be bidirectional, as a String may cast to an Integer, and an Integer may cast to a String.
+func (pt ParameterType) String() string {
+	switch pt {
+	case ParameterType_Null:
+		return "null"
+	case ParameterType_Integer:
+		return "integer"
+	case ParameterType_Float:
+		return "double precision"
+	case ParameterType_Numeric:
+		return "numeric"
+	case ParameterType_String:
+		return "character varying"
+	case ParameterType_Timestamp:
+		return "timestamp"
+	default:
+		panic(fmt.Errorf("unhandled type in ParameterType::String (%d)", int(pt)))
+	}
+}
+
+// ParameterTypeFromReflection returns the ParameterType and equivalent sql.Type from the given reflection type. If the
+// given type does not match a ParameterType, then this returns false.
+func ParameterTypeFromReflection(t reflect.Type) (ParameterType, sql.Type, bool) {
+	switch t {
+	case reflect.TypeOf(Integer{}):
+		return ParameterType_Integer, types.Int64, true
+	case reflect.TypeOf(Float{}):
+		return ParameterType_Float, types.Float64, true
+	case reflect.TypeOf(Numeric{}):
+		//TODO: properly handle decimal types
+		return ParameterType_Numeric, types.Float64, true
+	case reflect.TypeOf(String{}):
+		return ParameterType_String, types.LongText, true
+	case reflect.TypeOf(Timestamp{}):
+		return ParameterType_Timestamp, types.Datetime, true
+	default:
+		return ParameterType_Null, types.Null, false
+	}
+}
+
+// IntermediateParameter is a parameter before it has been finalized.
+type IntermediateParameter struct {
+	Value        interface{}
```

Comment:
Here and in other places, we should probably use the more modern `any` alias instead of `interface{}`